File: C++/ShortestPath/2976 MinimumCostToConvertString.cpp
// LeetCode 2976. Minimum Cost to Convert String I
// Time Complexity: O(26 * (E log V)) ≈ O(E log V)
// Space Complexity: O(26 * 26 + E)

#include <vector>
#include <queue>
#include <string>
#include <climits>
using namespace std;

class Solution {
public:
    long long minimumCost(
        string source,
        string target,
        vector<char>& original,
        vector<char>& changed,
        vector<int>& cost) {

        vector<vector<pair<int,int>>> adj(26);

        for (int i = 0; i < original.size(); i++) {
            adj[original[i] - 'a'].push_back(
                {changed[i] - 'a', cost[i]}
            );
        }

        vector<vector<long long>> dist(26, vector<long long>(26, -1));
        vector<bool> processed(26, false);

        // Run Dijkstra once per unique source character
        for (char c : source) {
            int idx = c - 'a';
            if (!processed[idx]) {
                dijkstra(idx, adj, dist[idx]);
                processed[idx] = true;
            }
        }

        long long total = 0;

        for (int i = 0; i < source.size(); i++) {
            int s = source[i] - 'a';
            int t = target[i] - 'a';

            if (s == t) continue;

            if (dist[s][t] == -1) return -1;
            total += dist[s][t];
        }

        return total;
    }

private:
    void dijkstra(
        int src,
        vector<vector<pair<int,int>>>& adj,
        vector<long long>& dist) {

        for (int i = 0; i < 26; i++) dist[i] = LLONG_MAX;
        dist[src] = 0;

        priority_queue<
            pair<long long,int>,
            vector<pair<long long,int>>,
            greater<>
        > pq;

        pq.push({0, src});

        while (!pq.empty()) {
            auto [d, u] = pq.top();
            pq.pop();

            if (d > dist[u]) continue;

            for (auto& [v, w] : adj[u]) {
                if (dist[v] > d + w) {
                    dist[v] = d + w;
                    pq.push({dist[v], v});
                }
            }
        }

        for (int i = 0; i < 26; i++) {
            if (dist[i] == LLONG_MAX) dist[i] = -1;
        }
    }
};

File: Java/All/2976 MinimumCostToConvertString.java
// LeetCode 2976. Minimum Cost to Convert String I
// Time Complexity: O(26 * (E log V)) ≈ O(E log V)
// Space Complexity: O(26 * 26 + E)

import java.util.*;

class Solution {

    class Pair {
        char node;
        long dist;

        Pair(char n, long d) {
            node = n;
            dist = d;
        }
    }

    public long minimumCost(
            String source,
            String target,
            char[] original,
            char[] changed,
            int[] cost) {

        int n = source.length();

        // Build directed graph
        Map<Character, List<Pair>> adj = new HashMap<>();
        for (int i = 0; i < original.length; i++) {
            adj.computeIfAbsent(original[i], k -> new ArrayList<>())
                    .add(new Pair(changed[i], cost[i]));
        }

        // dist[u][v] = minimum cost to convert u -> v
        long[][] dist = new long[26][26];
        for (int i = 0; i < 26; i++) {
            Arrays.fill(dist[i], -1);
        }

        boolean[] processed = new boolean[26];

        // Run Dijkstra once per unique source character
        for (char c : source.toCharArray()) {
            int idx = c - 'a';
            if (!processed[idx]) {
                dijkstraAll(c, adj, dist[idx]);
                processed[idx] = true;
            }
        }

        long total = 0;

        for (int i = 0; i < n; i++) {
            char s = source.charAt(i);
            char t = target.charAt(i);

            if (s == t) continue;

            long d = dist[s - 'a'][t - 'a'];
            if (d == -1) return -1;

            total += d;
        }

        return total;
    }

    // Dijkstra from a single source to all characters
    private void dijkstraAll(
            char src,
            Map<Character, List<Pair>> adj,
            long[] dist) {

        Arrays.fill(dist, Long.MAX_VALUE);
        dist[src - 'a'] = 0;

        PriorityQueue<Pair> pq =
                new PriorityQueue<>((a, b) -> Long.compare(a.dist, b.dist));
        pq.offer(new Pair(src, 0));

        while (!pq.isEmpty()) {
            Pair cur = pq.poll();
            char u = cur.node;
            long d = cur.dist;

            if (d > dist[u - 'a']) continue;
            if (!adj.containsKey(u)) continue;

            for (Pair p : adj.get(u)) {
                char v = p.node;
                long nd = d + p.dist;

                if (nd < dist[v - 'a']) {
                    dist[v - 'a'] = nd;
                    pq.offer(new Pair(v, nd));
                }
            }
        }

        // Mark unreachable nodes
        for (int i = 0; i < 26; i++) {
            if (dist[i] == Long.MAX_VALUE) {
                dist[i] = -1;
            }
        }
    }
}
